app/controllers/fhir/r4/api_controller.rb

```diff
@@ -426,6 +426,132 @@ def all
+    # Add the job_id of the job to the download for reference in status requests
+    download.update(job_id: export_job.provider_job_id)
+
+    respond_to do |format|
+      format.any { head(:accepted, content_location: "#{root_url}fhir/r4/ExportStatus/#{download.id}") }
+    end
+  end
+
+  # Give an update on the status of a bulk FHIR data request
+  #
+  # GET /fhir/r4/ExportStatus/[:id]
+  def export_status
+    id = params.require(:id)
+    download = ApiDownload.where(application_id: doorkeeper_token&.application_id).find_by_id(id)
+    status = Sidekiq::Status.status(download&.job_id)
+    case status
+    when :complete
+      begin
+        transaction_time = DateTime.parse(Sidekiq::Status.get(download.job_id, :transaction_time))
+      rescue Date::Error
+        status_server_error(['Export failed']) && return
+      end
+
+      response_json = {
+        transactionTime: transaction_time.utc.strftime('%FT%T%:z'),
+        request: download.url,
+        requiresAccessToken: true,
+        errors: [],
+        output: download.files.blobs.pluck(:filename).map do |file|
+          resource_type = file.split('.').first
+          {
+            type: resource_type,
+            url: "#{root_url}fhir/r4/ExportFiles/#{id}/#{resource_type}"
+          }
+        end
+      }
+      render json: response_json.to_json, status: :ok
+    when :failed
+      status_server_error(['Export failed']) && return
+    when :working, :queued, :retrying
+      response.headers['X-Progress'] = if status == :working
+                                         "#{Sidekiq::Status.pct_complete(download.job_id)} %"
+                                       else
+                                         status.to_s
+                                       end
+      response['Retry-After'] = Rails.configuration.api['bulk_export_status_retry_after_seconds'].to_s
+      head :accepted
+    else
+      status_not_found_with_custom_errors(['No export found for this ID']) && return
+    end
+  end
+
+  # Return an bulk exported ndjson file
+  #
+  # GET /ExportFiles/[:id]/[:resource_type]
+  def export_files
+    unless request.headers['Accept'].nil? || request.headers['Accept'] == 'application/fhir+ndjson'
+      status_not_acceptable_with_custom_errors(["'Accept' header must have a value of 'application/fhir+ndjson'"]) && return
+    end
+
+    id = params.require(:id)
+    resource_type = params.require(:resource_type)&.downcase
+    authorize_resource_type_read(resource_type)
+
+    download = current_client_application&.api_downloads&.find_by_id(id)
+    status_not_found_with_custom_errors(['No file found at this URL']) && return if download.nil?
+
+    # Set the headers before streaming the returned content
+    response.headers['Content-Type'] = 'application/fhir+ndjson'
+    # Find and return the specific attachment associated with the given filename
+    download.files.blobs.where(filename: "#{resource_type}.ndjson").first.download do |chunk|
```

Comment:
This should have error handling for if `download.files.blobs.where(filename: "#{resource_type}.ndjson").first` is `nil`. Right now you get `undefined method `download' for nil:NilClass`, resulting in a 500 error. I guess a 404 would make the most sense in this scenario.

Reply:
Addressed in https://github.com/SaraAlert/SaraAlert/pull/987/commits/fef5e9e9487bf3f37b213141467c2b2ae42fff10.